### ISSUE TITLE
twisted.web.http.Request.parseCookies has incomplete test coverage

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6119]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6119|
|Type|defect|
|Created|2012-10-16 13:08:55Z|
|Branch|https://github.com/twisted/twisted/tree/parsecookies-coverage-6119|

It should have complete test coverage (found while porting http.py to Python 3).

Attachments:

* [6119.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3cd/3cd7a38a436f8c2139668b4743f20a57ba90d263/27fa66bccc7517110d3e799f84d22b97ccf6e245.patch) (685 bytes) - added by moijes12 on 2013-03-29 11:19:55Z - 
* [6119.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3cd/3cd7a38a436f8c2139668b4743f20a57ba90d263/49ecda58e92a4dd3d9cdd9626641987265688850.patch) (1567 bytes) - added by moijes12 on 2013-07-14 13:41:36Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6119 6119
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_parsecookies_coverage_6119 branches/parsecookies-coverage-6119
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1350392935000000 1350392935000000
changetime__1378735926000000 1378735926000000
version__None None
owner__lvh lvh
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_2"></a>|@moijes12 commented|
|-|-|

Hi

As far as I can see, there are tests for verifying that http.Request parses cookies in the below ways:
 1. Extract cookies from header and add them to recieved_cookies.
 2. Extract cookies from multiple cookie headers.

I think there is not much to add here except add a check to see that parseCookies function returns None in case there is no cookie header. This is implemented by the below line in parseCookie
     
   if cookieheaders is None:
            return

I have added the test for the above check. May be we could check that recieved_cookies is empty if the cookie header was not contained in the request. We could implement that if required.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince set owner to @moijes12|
|-|-|

1.`parseCookies` doesn't have an interesting return value, so there is no reason to assert anything about it. But you are right that that code path isn't covered. If there isn't any headers, then no cookies should be parsed.
 2. The `except ValueError:` is also not covered. You should test that malformed cookie-pair's are ignored (and other cookies set before or after are still properly parsed.

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_4"></a>|@moijes12 removed owner|
|-|-|

I am attaching a patch which has the test covering the "ValueError" codepath i.e. it tests that malformed-cookie pair's are ignored. Here I have added 2 tests in a single test function.   

1.  Test that malformed cookie-pair before, in middle or at end of header are ignored.  

2. Test that malformed URLs in multiple headers cookies are ignored as in 1. 

 I have used the code in test_parseCookie and test_parseCookiesMultipleHeaders for achieving this.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj commented|
|-|-|

(In [39650]) Branching to 'parsecookies-coverage-6119'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj commented|
|-|-|

(In [39651]) Apply 6119.2.patch from moijes12. Refs: [#6119](https://github.com/twisted/twisted/issues/6119)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj set owner to @moijes12|
|-|-|

Thanks moijes12,

Notes:
 * Merges cleanly
 * Coverage now seems complete

Points:
 1. Build Failures
    1. Trailing whitespace:
       https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/1150/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
       * Fixed r39666
 2. [branches/parsecookies-coverage-6119/twisted/web/test/test_http.py](branches/parsecookies-coverage-6119/twisted/web/test/test_http.py)
    1. Should be two blank lines before "def test_parseCookiesNoCookie(self):"
       * Fixed r39666
    2. Builtin types such as None should be marked up with L{}
       * Fixed r39667
    3. test_parseCookiesNoCookie tests for a None return value, but
       the function always returns None. So perhaps there should be separate
       tests for return type and another test that
       req.received_cookies is empty if no cookies are supplied.
       * Fixed in r39668
    4. test_parseCookiesMalformedCookie should probably be split into
       two separate tests.
       * Fixed in r39669
    5. parseCookies does an lstrip on the value so add a test for
       leaving right hand space in the cookie value.
       * Fixed in r39670
    6. And a test for removing leading space from cookie key
       * Fixed in r39672
    7. parseCookies allows cookies with empty values. Add a test for that.
       * Fixed in r39671
 3. parseCookie seems inadequate in various ways:
    1. Allows empty keys " = "
    2. Allows spaces in values "foo = bar baz; "
    3. Misinterprets quoted keys '"foo=bar";baz=qux'
    4. Raise a ticket
       * Created [#6694](https://github.com/twisted/twisted/issues/6694)

Ok. This started as a code review but I also fixed some things as a
went.

So I'll force a new build and put it back up for review.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj removed owner|
|-|-|

Back up for review. 

Build Results:
 * http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/parsecookies-coverage-6119&num_builds=2

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_9"></a>|@lvh set owner to @lvh|
|-|-|

Hi rwall,

Thanks for working on this!

The twistedchecker buildbot noted an issue:

************* Module twisted.web.test.test_http
W9010:1486,0: Trailing whitespace found in the end of line

Other than that, these tests look very reasonable, modulo the already existing bogus behavior which is addressed in another ticket.

Since these are two very minor changes, I'll do them.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_10"></a>|@lvh set status to `closed`|
|-|-|

(In [39897]) 
Merge parsecookies-coverage-6119: add tests for cookie parsing

Author: rwall, moijes12
Reviewer: lvh, rwall, tom.prince
Fixes: [#6119](https://github.com/twisted/twisted/issues/6119)

The cookie parsing logic previously had no test coverage. Now it does.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

The branch is also introducing tests that pass `None` as a false value for a flag parameter (`Request.__init__` `queued`).  This works but is weird.  The code should pass `False` instead.
